x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -530,6 +532,8 @@ protected Clock getClock() {
             dlsBitsetCache.get(), new DeprecationRoleDescriptorConsumer(clusterService, threadPool));
         securityIndex.get().addIndexStateListener(allRolesStore::onSecurityIndexStateChange);
 
+        components.add(new ApiKeyGenerator(apiKeyService, allRolesStore, clusterService, xContentRegistry));
```

Comment:
This and relevant changes will be something to "forward-port"

Reply:
Why would we add this to `components`?
We're trying to de-guice, and this seems like a step in the wrong direction.

Reply:
Both `TransportGrantApiKeyAction` and `TransportCreateApiKeyAction` needs it. So it is a deduplication. Also because the current changes require `clusterService` to be available for `ApiKeyGenerator`, so instead of inject `clusterService` into both the acton classes, it is now passed only to the thing that really needs it. The same argument can be applied for `apiKeyService` and `allRolesStore` as well. 
Yes de-guice is what we want, but in the meantime, all Action classes (old and new) still need `@Inject`. I think it is ok to have another one and it will just be batch de-guiced with all the existing ones when the time comes

Reply:
Nevermind. After moving the validation logic from `ApiKeyGenerator` to `ApiKeyService`, there is no point to touch `ApiKeyGenerator` anymore. I will revert related changes.